File: Python/HackerRank_Practice_Lab/Mathematical_Operations/nonDivisibleSubset.py

from collections import Counter

def non_divisible_subset(k, s):
    """
    Calculates the size of the largest subset of 's' such that the sum of 
    any two integers in the subset is not divisible by 'k'.

    Args:
        k (int): The integer divisor.
        s (list[int]): An array of integers.

    Returns:
        int: The size of the maximum non-divisible subset.
    """
    remainder_count = Counter(x % k for x in s)
    result = 0

    if remainder_count[0] > 0:
        result += 1

    for r in range(1, (k // 2) + 1):
        if r == k - r:
            if remainder_count[r] > 0:
                result += 1
        else:
            result += max(remainder_count[r], remainder_count[k - r])

    return result

def main():
    """
    Reads input from stdin, parses the values, and prints the result of 
    the non_divisible_subset function.
    """
    first_multiple_input = input().rstrip().split()

    n = int(first_multiple_input[0])

    k = int(first_multiple_input[1])

    s = list(map(int, input().rstrip().split()))

    if n != len(s):
        raise ValueError("The length of s should be equal to n")

    print(non_divisible_subset(k, s))

if __name__ == "__main__":
    main()

File: problems/hackerrank/math/nonDivisibleSubset/nonDivisibleSubset.go
package math



// NonDivisibleSubset calculates the size of the largest subset where sum of any two is not divisible by k.
func NonDivisibleSubset(k int, s []int) int {
	remainderCount := make(map[int]int)
	for _, x := range s {
		remainderCount[x%k]++
	}

	result := 0
	if remainderCount[0] > 0 {
		result += 1
	}

	for r := 1; r <= k/2; r++ {
		if r == k-r {
			if remainderCount[r] > 0 {
				result += 1
			}
		} else {
			c1 := remainderCount[r]
			c2 := remainderCount[k-r]
			if c1 > c2 {
				result += c1
			} else {
				result += c2
			}
		}
	}

	return result
}
